accord-core/src/main/java/accord/coordinate/tracking/AbstractTracker.java

```diff
@@ -0,0 +1,205 @@
+
+        @Override
+        public ShardOutcomes apply(AbstractTracker<?, ?> tracker)
+        {
+            if (this == Success)
+                --tracker.waitingOnShards;
+            return this;
+        }
+    }
+
+    final Topologies topologies;
+    protected final ST[] trackers;
+    protected int waitingOnShards;
+
+    AbstractTracker(Topologies topologies, IntFunction<ST[]> arrayFactory, Function<Shard, ST> trackerFactory)
+    {
+        this(topologies, arrayFactory);
+        int topologyCount = topologies.size();
+        int maxShards = maxShardsPerEpoch();
+        int waitingOnShards = 0;
+        for (int i = 0 ; i < topologyCount ; ++i)
+        {
+            Topology topology = topologies.get(i);
+            int size = topology.size();
+            for (int j = 0; j < size; ++j)
+                trackers[i * maxShards + j] = trackerFactory.apply(topology.get(j));
+            waitingOnShards += topologies.get(i).size();
+        }
+        this.waitingOnShards = waitingOnShards;
+    }
+
+    AbstractTracker(Topologies topologies, IntFunction<ST[]> arrayFactory)
+    {
+        Preconditions.checkArgument(topologies.totalShards() > 0);
+        int topologyCount = topologies.size();
+        int maxShards = topologies.get(0).size();
+        for (int i = 1 ; i < topologyCount ; ++i)
+            maxShards = Math.max(maxShards, topologies.get(i).size());
+        this.topologies = topologies;
+        this.trackers = arrayFactory.apply(topologyCount * maxShards);
+    }
+
+    protected int topologyOffset(int topologyIdx)
+    {
+        return topologyIdx * maxShardsPerEpoch();
+    }
+
+    public Topologies topologies()
+    {
+        return topologies;
+    }
+
+    protected RequestStatus trySendMore() { throw new UnsupportedOperationException(); }
+
+    <T extends AbstractTracker<ST, P>>
+    RequestStatus recordResponse(T self, Id node, BiFunction<? super ST, P, ? extends ShardOutcome<? super T>> function, P param)
+    {
+        return recordResponse(self, node, function, param, topologies.size());
+    }
+
+    <T extends AbstractTracker<ST, P>>
+    RequestStatus recordResponse(T self, Id node, BiFunction<? super ST, P, ? extends ShardOutcome<? super T>> function, P param, int topologyLimit)
+    {
+        Preconditions.checkState(self == this); // we just accept self as parameter for type safety
+        ShardOutcomes minShardStatus = ShardOutcomes.NoChange;
+        int maxShards = maxShardsPerEpoch();
+        for (int i = 0; i < topologyLimit && minShardStatus != ShardOutcomes.Fail; ++i)
+        {
+            minShardStatus = topologies.get(i).mapReduceOn(node, i * maxShards, AbstractTracker::apply, self, function, param, ShardOutcomes::min, minShardStatus);
+        }
+
+        switch (minShardStatus)
+        {
+            default: throw new AssertionError();
+            case SendMore:
+                return trySendMore();
+            case Success:
+                if (waitingOnShards == 0)
```

Comment:
This `Success` is per-shard, i.e. it means at least one shard returned success, and no shards returned failure. So if only one shard has returned `Success` but not all, we could absolutely still have `waitingOnShards != 0`

Reply:
Ah, I keep getting caught by that distinction when we do the min check across something that is over multiple shard and all it means is one shard has achieved that state.

Reply:
No worries. Honestly I’ve vacillated over this a lot. I think it might be cleaner to return a global result from each shard’s AcceptOutcomes (since it can compute it) but I didn’t want to pollute the RequestOutcome with TryMore. I’m far from sold I got the balance right, I think it would be nicer to encapsulate the decision in AcceptOutcomes so the min is computed over the whole tracker. We could introduce an intermediate enum TrackerStatus that includes TryMore, I think that would be enough to clean this up a bit more.